### ISSUE TITLE
EXIF date/time not correctly parsed

### ISSUE THREAD

**Issue by uvok** [darktable-org/darktable#17389]
### Describe the bug

When I import *some* images taken with my smartphone camera (hardware: Pixel 6a, software: OpenCamera), Darktable shows 01/01/01 as datetime. This seems to be related to the way the EXIF data is "formatted".

Other images, taken on the same phone, with the same app, work fine. (Don't ask me why! I have no idea.) Not sure if I should report a bug to OpenCamera as well due to their EXIF data formatting?

This was once already fixed, apparently: #13355

### Steps to reproduce

1. Download the image https://nc.uvokchee.de/s/BXtq9Dery2x24cT
2. Import it in Darktable
3. DateTime image information now displays the datetime as 01/01/01

### Expected behavior

Darktable should parse the EXIF data correctly, show the datetime correctly

### Logfile | Screenshot | Screencast

exiftool output of "broken" image:

```
$ exiftool 'IMG_20240813_171426.jpg'  | grep 2024
File Name                       : IMG_20240813_171426.jpg
Directory                       : /media/pictures/2024_EAST_12
File Modification Date/Time     : 2024:08:13 17:14:26+02:00
File Access Date/Time           : 2024:08:24 18:43:51+02:00
File Inode Change Date/Time     : 2024:08:24 18:47:31+02:00
Modify Date                     : 2024:08:13 17:14:26
Date/Time Original              : 2024:08:13 17:14:26
Create Date                     : 2024:08:13 17:14:26
Create Date                     : 2024:08:13 17:14:26.18261+02:00
Date/Time Original              : 2024:08:13 17:14:26.18261+02:00
Modify Date                     : 2024:08:13 17:14:26.18261+02:00
```

exiftool output of "correctly parsed image":

```
$ exiftool 'IMG_20240814_213346.jpg'  | grep 2024
File Name                       : IMG_20240814_213346.jpg
Directory                       : /media/pictures/2024_EAST_12
File Modification Date/Time     : 2024:08:14 21:33:46+02:00
File Access Date/Time           : 2024:08:24 18:43:51+02:00
File Inode Change Date/Time     : 2024:08:24 18:47:31+02:00
Modify Date                     : 2024:08:14 21:33:46
Date/Time Original              : 2024:08:14 21:33:46
Create Date                     : 2024:08:14 21:33:46
Create Date                     : 2024:08:14 21:33:46.9275
Date/Time Original              : 2024:08:14 21:33:46.9275
Modify Date                     : 2024:08:14 21:33:46.9275
```

The timezone specification seems to throw off Darktable

### Where did you obtain darktable from?

downloaded from www.darktable.org

### darktable version

4.8.1

### What OS are you using?

Linux

### What is the version of your OS?

Devuan Daedelus (eq. to Debian Bookworm)

### Describe your system?

Language/Localization is set to German! Maybe that matters as well.

### Are you using OpenCL GPU in darktable?

None

**Comment by uvok** [comment]
Version used: AppImage, Flatpak. - Same behavior.

**Comment by ptilopteri** [comment]
* uvok cheetah ***@***.***> [08-28-24 13:47]:
> Version used: AppImage, Flatpak. - Same behavior.

jfyi, "appimage" and "flatpak" are not version but distributions/packages. 
Versions have numbers, current master is "release-4.9.0-326-g28f3414f3c"

 
-- 
(paka)Patrick Shanahan       Plainfield, Indiana, USA          @ptilopteri
facebook/ptilopteri
Photos: http://wahoo.no-ip.org/piwigo                   paka @ IRCnet oftc

**Comment by uvok** [comment]
For completeness sake:

```
$ exiv2 -pa 'IMG_20240813_171426.jpg'  | grep Date
Exif.Image.DateTime                          Ascii      21  2024:08:13 17:14:26
Exif.Photo.DateTimeOriginal                  Ascii      21  2024:08:13 17:14:26
Exif.Photo.DateTimeDigitized                 Ascii      21  2024:08:13 17:14:26
$ exiv2 -pa 'IMG_20240814_213346.jpg'  | grep Date
Exif.Image.DateTime                          Ascii      20  2024:08:14 21:33:46
Exif.Photo.DateTimeOriginal                  Ascii      20  2024:08:14 21:33:46
Exif.Photo.DateTimeDigitized                 Ascii      20  2024:08:14 21:33:46
```

Also, I tried building some versions myself and re-tested the behavior.

Tag release-4.4.2: Does not work either
Tag release-4.2.1: Does not work either

Also, maybe related to #14244

(I deleted the sidecar files before each test)

**Comment by uvok** [comment]
Did some debugging on current master.

`_find_datetime_taken` apparently doesn't find the exif tag, jumps to the else branch (`*exif_datetime_taken = '\0')

(Need to install the exiv2 debugging symbols as well to dig further, I guess)

**Comment by zisoft** [comment]
`_find_datetime_taken()` checks for a valid length specifier `DT_DATETIME_EXIF_LENGTH` which is `20`.

The "good" photo has a correct length (see the 20 in the output):
```
$ exiv2 -pa 'IMG_20240814_213346.jpg'  | grep Date
Exif.Photo.DateTimeOriginal                  Ascii      20  2024:08:14 21:33:46
```
The "bad" photo indicates a 21 here:
```
$ exiv2 -pa 'IMG_20240813_171426.jpg'  | grep Date
Exif.Photo.DateTimeOriginal                  Ascii      21  2024:08:13 17:14:26
```

**Comment by zisoft** [comment]
```
  | | 4)  DateTimeOriginal = 2024:08:13 17:14:26
  | |     - Tag 0x9003 (21 bytes, string[21]):
  | |         017a: 32 30 32 34 3a 30 38 3a 31 33 20 31 37 3a 31 34 [2024:08:13 17:14]
  | |         018a: 3a 32 36 00 00                                  [:26..]
```
We have two nul termination bytes here.

**Comment by zisoft** [comment]
https://github.com/darktable-org/darktable/blob/a03a2d1adc39123ff4d2d637311500f4a2bf1ff0/src/common/exif.cc#L1088-L1095

A simple `>=` would work. It still will copy only the first 20 characters.

`&& pos->size() >= DT_DATETIME_EXIF_LENGTH) `

But this is only a workaround because the spec is violated in that file.

**Comment by zisoft** [comment]
@victoryforce: you have read my mind :)

**Comment by victoryforce** [comment]
> But this is only a workaround because the spec is violated in that file.

The spec is violated, but the image contains all the information and we can read it. Why is it a workaround then? This is simply the correct processing of data that we are able to process correctly.

> Not sure if we should deal with an invalid file in such a way.

In what sense is this file invalid?

**Comment by victoryforce** [comment]
> @victoryforce: you have read my mind :)

Great minds think alike! 😊

**Comment by uvok** [comment]
Thanks, works fine for newly imported images. I will also raise an issue at OpenCamera.

Is there a way I can make Darktable re-read the EXIF metadata on the JPEG files already in the Darktable database (which also already have an XMP file with `exif:DateTimeOriginal=""`), without losing ratings? Or probably the easiest way is to do so via exiftool https://exiftool.org/metafiles.html

**Comment by zisoft** [comment]
"actions on selection" module --> "metadata" tab --> refresh EXIF

![Bildschirmfoto 2024-08-29 um 19 15 00](https://github.com/user-attachments/assets/5e793625-a264-4717-9362-60a2d03820d3)

Be sure to follow the instructions on the tooltip to keep existing ratings: Set "ignore EXIF rating" in the import module first:

![Bildschirmfoto 2024-08-29 um 19 15 21](https://github.com/user-attachments/assets/d11cb791-509d-4063-b026-77a16c1ea5b1)

**Comment by uvok** [comment]
Thanks! I didn't see the metadata tab (that it was a tab at all).

**Comment by victoryforce** [comment]
> > In what sense is this file invalid?
> 
> When something is out of spec, all bets are off. Might get lucky like this time, might not...

Yes, this is exactly my idea: out of spec data is not necessarily invalid data.

It is fundamentally important to always adhere to the principle: the interpretation of the spec should be as strict as possible on the side of the data creator and as permissive as possible on the side of the data reader.
